### ISSUE TITLE
Adaptions for n_classes>2

### ISSUE THREAD

**Issue by Murmeltier105** [choosehappy/PytorchDigitalPathology#13]
Thank you for sharing this sophisticated script. 
I got problems figuring out which exact changes to make in order to run it with data containing 7 classes of which I want to ignore the one containing background: first while making th hdf5 file and second during the training. 
Is it possible you could explain the corresponding lines in detail?

**Comment by choosehappy** [comment]
Should be pretty straightforward

all you need to do it set the model to accept more classes here:

https://github.com/choosehappy/PytorchDigitalPathology/blob/7edb9a4f53f6b289043b57bb003f5c3ce658c112/segmentation_epistroma_unet/train_unet_albumentations.py#L103

and modify the cross-entropy similarly here:

https://github.com/choosehappy/PytorchDigitalPathology/blob/7edb9a4f53f6b289043b57bb003f5c3ce658c112/segmentation_epistroma_unet/train_unet_albumentations.py#L207

Look closer at the cross entropy loss:

https://pytorch.org/docs/master/generated/torch.nn.CrossEntropyLoss.html

you'll see you can set the ignore_index to anything you want to ignore the
associated background

**Comment by Murmeltier105** [comment]
Thank you for the quick response.
The major problem seemed to be with the masks: transparency and variable size at the wsi boarder.
No it only mentions:
`/.local/lib/python3.6/site-packages/torch/nn/_reduction.py:46: UserWarning: size_average and reduce args will be deprecated, please use reduction='none' instead.
  warnings.warn(warning.format(ret))
/PytorchDigitalPathology/segmentation_epistroma_unet/train_EndometriumUNET.py:302: RuntimeWarning: invalid value encountered in true_divide
  all_acc[phase] = (cmatrix[phase] / cmatrix[phase].sum()).trace()`

**Comment by choosehappy** [comment]
this looks like an issue with the confusion matrix calculation

it may in fact "compress" (think squeeze) if all the classes aren't
present, resulting in mismatched sizes

it looks like you can simply comment out associated code if you don't care
about the measuring the accuraries in real time, otherwise it may need to
be refactored a bit

**Comment by Murmeltier105** [comment]
Having a closer look at it, I think there are more lines to be modified for multiple classes:

- in **make_hdf5**:  when reading the mask with `io=cv2.imread(fname)/255`  the division by 255 should be omitted as for classes >2 there's no binary nature

- in **train_unet**: inside the definition of Dataset, I reduced the weight definition to  `weight = scipy.ndimage.morphology.binary_dilation(mask, iterations=2) & ~mask` isntead of `[...] mask == 1` though I doubt, that binary dilation will yield useful results for a non binary mask. I'm currently out of plans what should be done here.

Combined with the visualization of a single example modified to 
`(img, patch_mask, patch_mask_weight) = dataset["train"][250]
fig, ax = plt.subplots(1, 4, figsize=(10, 4))  # 1 row, 2 columns

build output showing original patch  (after augmentation), class = 1 mask, weighting mask, overall mask (to see any ignored classes)
ax[0].imshow(np.moveaxis(img.numpy(), 0, -1))
ax[0].set_title('augmented Image')

ax[1].imshow(patch_mask) #instead of patch_mask == 1
ax[1].set_title('mask')

ax[2].imshow(patch_mask_weight)
ax[2].set_title('weighting mask')


ax[3].imshow(patch_mask)
ax[3].set_title('overall mask')
plt.show()`

I got the  figure:
![1000-CheckUp](https://user-images.githubusercontent.com/59916362/90124941-756f0e00-dd61-11ea-88a6-b8dd743a2dfe.png)

I changed the definition of the cmatrix to `cmatrix = {key: np.zeros((nclasses, nclasses)) for key in phases}`

At the moment that yields an error while trying to compute
`loss_matrix = criterion(prediction, y)` 
`Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)` when computing on cpu 
or `illegal memory access was encountered` when computing with cuda

**Comment by choosehappy** [comment]
> * in **make_hdf5**:  when reading the mask with `io=cv2.imread(fname)/255`  the division by 255 should be omitted as for classes >2 there's no binary nature

yup. correct. you want to save the mask images as uint8, with each pixel being [0,1,2,Nclass,Nclass+1]. where Nclass+1 will be your "ignore" class.  note that you cannot use e.g., -100 as uint8 is not signed

> * in **train_unet**: inside the definition of Dataset, I reduced the weight definition to  `weight = scipy.ndimage.morphology.binary_dilation(mask, iterations=2) & ~mask` isntead of `[...] mask == 1` though I doubt, that binary dilation will yield useful results for a non binary mask. I'm currently out of plans what should be done here.

there is no requirement to have this, so if it makes your life easier you can comment it and associated other parts out. it is pretty specific for segmentation cases which has pronoucned boundaries, so if one of your classes in the future struggles, you can come back to this, extract that particular class's mask, do the dilation, and save it to further boost the weighting in the loss function for those pixels


> Combined with the visualization of a single example modified to
> `(img, patch_mask, patch_mask_weight) = dataset["train"][250]
> fig, ax = plt.subplots(1, 4, figsize=(10, 4)) # 1 row, 2 columns
> 
> build output showing original patch (after augmentation), class = 1 mask, weighting mask, overall mask (to see any ignored classes)
 
> I got the figure:
> ![1000-CheckUp](https://user-images.githubusercontent.com/59916362/90124941-756f0e00-dd61-11ea-88a6-b8dd743a2dfe.png)

how many classes did you set? this mask image you're showing has 3 classes, purple, yellow, and blue

those classes need to be of values [0,1,2]. i would use np.unique on the mask to ensure that this true

> At the moment that yields an error while trying to compute
> `loss_matrix = criterion(prediction, y)`
> `Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)` when computing on cpu
> or `illegal memory access was encountered` when computing with cuda

this error is likely because of what i mentioned above. look at the unique values in y. remember this is going to do one hot encoding in the background for you. look closely at the pytorch documentation and see what the expected input is, both in terms of data size + data types

**Comment by Murmeltier105** [comment]
Thank you very much for your enduring support. 

I got 6 labels plus background. But labelled regions in the wsi are quite big so there are plenty of patches with only one label.
After your proposed adjustments the major problem seems to be that for the hdf5 file pictures are opened with cv2 which yields RGB values. These are anything but in the range(0,6) though the values for p mode pngs fit inside the range with Image.open.

I got it going with this workaround in make_hdf5 line 95:
`imgArr = np.asarray(Image.open(fname))`
` io = np.zeros((imgArr.shape[0],imgArr.shape[1],3))`
` io[:,:,0] = io[:,:,1] = io[:,:,2] = imgArr`
